### PR TITLE
Tomioka giyu

### PR DESCRIPTION
**Pull request by HiNxPeR**
Take 2 for Tomioka Giyu

### CODE REVIEW

File: src/mechanics/ingame/demonslayer/giyu/TomiokaGiyu.java [hunk 1]
```diff
@@ -1,37 +1,82 @@
...
+	private int instantKillBoostFromAbility2;
 
 	public TomiokaGiyu(int id) {
 		super(id);
-		// TODO Auto-generated constructor stub
+
+		this.setA1(new DeadCalm(this.getAbilities()[0]));
+		this.setA2(new SurfaceSlash(this.getAbilities()[1]));
+		this.setA3(new FinalBeheading(this.getAbilities()[2]));
+		this.setA4(new DefensiveStance(this.getAbilities()[3]));
```
Comment (acreunited) [inline src/mechanics/ingame/demonslayer/giyu/TomiokaGiyu.java:18]: Setting the Abilities here on the contructor is no longer necessary.

Theres no problem in being here, Im just letting you know that you dont need to do this anymore :)

File: src/mechanics/ingame/demonslayer/giyu/TomiokaGiyu.java [hunk 2]
```diff
@@ -1,37 +1,82 @@
...
 
 	@Override
-	public void applyAbility1(Ability a, Character c) {
-		// TODO Auto-generated method stub
+	public void applyAbility1(Ability ability, Character target) {
+		if (this.instantKillBoostFromAbility1 == 0) {
+			String[] abilityText = {"Water Breathing 11th Form: Dead Calm is improved"};
+			String[] abilityDuration = {Texts.INFINITE};
+
+			target.addActiveSkill(this, ability, abilityText, abilityDuration, false);
```
Comment (acreunited) [inline src/mechanics/ingame/demonslayer/giyu/TomiokaGiyu.java:30]: Here, the target in the argument is the enemy Character. 

The active skill should be added on our Character (Giyu), like this:
```suggestion
			this.addActiveSkill(this, ability, abilityText, abilityDuration, false);	
```

File: src/mechanics/ingame/demonslayer/giyu/TomiokaGiyu.java [hunk 3]
```diff
@@ -1,37 +1,82 @@
...
 	@Override
-	public void applyAbility2(Ability a, Character c) {
-		// TODO Auto-generated method stub
-		
+	public void applyAbility2(Ability ability, Character target) {
+		if (this.instantKillBoostFromAbility2 == 0) {
+			String[] abilityText = {"Water Breathing 1st Form: Water Surface Slash is improved"};
+			String[] abilityDuration = {Texts.INFINITE};
+
+			target.addActiveSkill(this, ability, abilityText, abilityDuration, false);
```
Comment (acreunited) [inline src/mechanics/ingame/demonslayer/giyu/TomiokaGiyu.java:44]: Same thing here, in the applyAbility1()

**Review by acreunited — CHANGES_REQUESTED** [review]
Its almost done! :)

**Review by acreunited — APPROVED** [review]
Aproved Chuan :)

There is something I noticed now while testing the Character again, I'll explain it to you asap

File: src/mechanics/ingame/demonslayer/giyu/TomiokaGiyu.java [hunk 4]
```diff
@@ -1,37 +1,82 @@
...
+		this.instantKillBoostFromAbility1 = 0;
+		this.instantKillBoostFromAbility2 = 0;
 	}
 
 	@Override
-	public void applyAbility1(Ability a, Character c) {
-		// TODO Auto-generated method stub
+	public void applyAbility1(Ability ability, Character target) {
+		if (this.instantKillBoostFromAbility1 == 0) {
+			String[] abilityText = {"Water Breathing 11th Form: Dead Calm is improved"};
```
Comment (acreunited) [inline src/mechanics/ingame/demonslayer/giyu/TomiokaGiyu.java:30]: When you apply the 1st and 2nd skill, it should appear that the 3rd skill is improved, and not the 1st nor 2nd Ability. Theres already a method in Texts that does this:
```suggestion
			String[] abilityText = {Texts.SKILL_IMPROVED(this.getA3().getName())};
```

File: src/mechanics/ingame/demonslayer/giyu/TomiokaGiyu.java [hunk 5]
```diff
@@ -1,37 +1,82 @@
...
+		this.instantKillBoostFromAbility1 = 10;
 	}
 
 	@Override
-	public void applyAbility2(Ability a, Character c) {
-		// TODO Auto-generated method stub
-		
+	public void applyAbility2(Ability ability, Character target) {
+		if (this.instantKillBoostFromAbility2 == 0) {
+			String[] abilityText = {"Water Breathing 1st Form: Water Surface Slash is improved"};
```
Comment (acreunited) [inline src/mechanics/ingame/demonslayer/giyu/TomiokaGiyu.java:44]: Same thing here:
`String[] abilityText = {Texts.SKILL_IMPROVED(this.getA3().getName())};`

File: src/mechanics/ingame/demonslayer/giyu/TomiokaGiyu.java [hunk 6]
```diff
@@ -1,37 +1,82 @@
...
 		super(id);
-		// TODO Auto-generated constructor stub
+		
+		this.instantKillBoostFromAbility1 = 0;
+		this.instantKillBoostFromAbility2 = 0;
+	}
+	
+	private void clearInstantKillBuffs() {
+		this.removeActive(this.getA1());
+		this.removeActive(this.getA2());
```
Comment (acreunited) [inline src/mechanics/ingame/demonslayer/giyu/TomiokaGiyu.java:21]: This one is my fault, for reviewing it wrongly. 
When you apply the 3rd Ability, you remove the active of the 1st and 2nd. That is correct and should happen, but the way I suggested doesnt work for this situation.

After you apply an Ability, a new object is created. So imagine for this Character, you apply the 1st Ability. Then use the 3rd Ability. Because a new 1st Ability was created, you are not removing the 1st Ability that was used, but the new one instead (and the new one has nothing to be removed). 

So basically, its necessary to store the 1st & 2nd Ability that boosted the 3rd ability effect, and remove those abilities.

A suggestion of mine is to store it on a list, as an attribute of the class:
`List<Ability> instantKillBoostAbilities = new ArrayList<>();`

Then here, we loop the list and reset it:
```suggestion
		for (Ability ability : instantKillBoostAbilities) {
			this.removeActive(ability);
		}
		this.instantKillBoostAbilities.clear();
```

**Review by acreunited — APPROVED** [review]
Giyu will be perfect after this! :)

**Review by acreunited — APPROVED** [review]
Perfect ^^
